Source/WebCore/rendering/RenderObject.cpp

```diff
@@ -2653,10 +2676,21 @@ auto RenderObject::collectSelectionGeometriesInternal(const SimpleRange& range)
     bool containsDifferentWritingModes = false;
     bool hasLeftToRightText = false;
     bool hasRightToLeftText = false;
+    bool needsToSeparateBoundary = false;
+    SingleThreadWeakPtr<RenderObject> previousRenderer;
     for (Ref node : intersectingNodesWithDeprecatedZeroOffsetStartQuirk(range)) {
         CheckedPtr renderer = node->renderer();
+        if (!renderer)
+            continue;
+
+        if (currentNodeRequiresToSeparateBoundary(renderer.get()))
+            needsToSeparateBoundary = true;
+        else if (!needsToSeparateBoundary)
+            needsToSeparateBoundary = lastNodeRequiresToSeparateBoundary(previousRenderer.get(), renderer->previousSibling());
```

Comment:
Might be more succinct this way:

```suggestion
        needsToSeparateBoundary = currentNodeRequiresToSeparateBoundary(renderer.get()) || lastNodeRequiresToSeparateBoundary(previousRenderer.get(), renderer->previousSibling());
```